File: 22-dp1/1-intro/main.py
# Dynamic Programming Intro - Python example
# Compute Fibonacci numbers using memoization

def fib(n, memo={}):
    if n in memo:
        return memo[n]
    if n <= 1:
        return n
    memo[n] = fib(n-1, memo) + fib(n-2, memo)
    return memo[n]

if __name__ == "__main__":
    n = 10
    print(f"Fibonacci number at position {n} is {fib(n)}")

File: 22-dp1/1-intro/main.java
// Dynamic Programming Intro - Java example
// Compute Fibonacci numbers using memoization

import java.util.HashMap;

public class Main {
    static HashMap<Integer, Integer> memo = new HashMap<>();
    
    public static int fib(int n) {
        if (memo.containsKey(n)) return memo.get(n);
        if (n <= 1) return n;
        int result = fib(n-1) + fib(n-2);
        memo.put(n, result);
        return result;
    }
    
    public static void main(String[] args) {
        int n = 10;
        System.out.println("Fibonacci number at position " + n + " is " + fib(n));
    }
}
